src/Control/Carrier/ContainerRegistryApi.hs

```diff
@@ -0,0 +1,314 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Control.Carrier.ContainerRegistryApi (
+  runContainerRegistryApi,
+  ContainerRegistryApiC,
+) where
+
+import App.Fossa.Analyze (updateProgress)
+import Codec.Archive.Tar qualified as Tar
+import Conduit (
+  ConduitT,
+  PrimMonad,
+  awaitForever,
+  runConduit,
+  runResourceT,
+  sinkFile,
+  yield,
+  (.|),
+ )
+import Container.Docker.Manifest (manifestFilename)
+import Container.Docker.OciManifest (
+  NotSupportedManifestFmt (NotSupportedManifestFmt),
+  OciManifestV2,
+  blobEntries,
+  supportedManifestKinds,
+  toDockerManifest,
+ )
+import Container.Docker.OciManifestIndex (digestOf, supportedManifestIndexKinds)
+import Container.Docker.SourceParser (
+  RegistryImageSource (
+    RegistryImageSource,
+    platformArchitecture,
+    registryContainerRepositoryReference,
+    registryCred
+  ),
+  RepoDigest,
+  RepoReference (RepoReferenceDigest),
+ )
+import Control.Algebra (Algebra, Has)
+import Control.Carrier.AtomicCounter (runAtomicCounter)
+import Control.Carrier.ContainerRegistryApi.Authorization (applyBearAuth, getAuthToken, mkRequest)
+import Control.Carrier.ContainerRegistryApi.Common (
+  RegistryCtx (RegistryCtx),
+  fromResponse,
+  getContentType,
+  getToken,
+ )
+
+import Control.Carrier.Finally (runFinally)
+import Control.Carrier.Reader (ReaderC, ask, runReader)
+import Control.Carrier.Simple (SimpleC, interpret)
+import Control.Carrier.StickyLogger (runStickyLogger)
+import Control.Carrier.TaskPool (withTaskPool)
+import Control.Concurrent (getNumCapabilities)
+import Control.Concurrent.STM (newEmptyTMVarIO)
+import Control.Effect.ContainerRegistryApi (
+  ContainerRegistryApiF (ExportImage, GetImageManifest),
+ )
+import Control.Effect.Diagnostics (
+  Diagnostics,
+  context,
+  fatal,
+  fatalText,
+  fromEither,
+  fromMaybeText,
+ )
+import Control.Effect.Lift (Lift, sendIO)
+import Control.Effect.Reader (Reader)
+import Control.Effect.TaskPool (forkTask)
+import Control.Monad (when)
+import Data.Aeson (eitherDecode, encode)
+import Data.ByteString (ByteString, writeFile)
+import Data.ByteString.Lazy qualified as ByteStringLazy
+import Data.Conduit.Zlib (ungzip)
+import Data.Maybe (fromMaybe)
+import Data.String.Conversion (
+  LazyStrict (toStrict),
+  toString,
+  toText,
+ )
+import Data.Text (Text)
+import Effect.Logger (
+  Logger,
+  Pretty (pretty),
+  Severity (SevInfo),
+  logDebug,
+  logInfo,
+ )
+import Effect.ReadFS (ReadFS, listDir)
+import Network.HTTP.Client (
```

Comment:
Note: Using `http-client`, instead of `req` is not ideal, but `req` does not have an easy way of dealing with exceptions and yielding responses (without substantive manager modification) and a simple API for dealing with response headers; we can use `req'` primitive like we use `http`, but at that instance pros are onlt modern-uri and nicer decode/json wrapping. Also http-conduit nicely fits with http-client. 

If we think this is critical to address, I prefer to do this in subsequent PR, or as ticketed item later on (before release).